Mathlib/AlgebraicTopology/SimplicialObject/Coskeletal.lean

```diff
@@ -0,0 +1,116 @@
+    Nonempty (IsRightKanExtension X (𝟙 ((Truncated.inclusion n).op ⋙ X)))
+
+variable [X.IsCoskeletal n]
+
+/-- If `X` is `n`-cosketal, then `𝟙 ((Truncated.inclusion n).op ⋙ X)` defines a right Kan
+extension of `(Truncated.inclusion.op ⋙ X)` along `(Truncated.inclusion n).op`. -/
+theorem IsCoskeletal.isRightKanExtension :
+    IsRightKanExtension X (𝟙 ((Truncated.inclusion n).op ⋙ X)) :=
+  IsCoskeletal.nonempty_isRightKanExtension.some
+
+/-- If `X` is `n`-coskeletal, then `rightExtensionInclusion X n` is a terminal object in the
+category `RightExtension (Truncated.inclusion n).op (Truncated.inclusion.op ⋙ X)`. -/
+noncomputable def IsCoskeletal.isUniversalOfIsRightKanExtension :
+    (rightExtensionInclusion X n).IsUniversal := by
+  have := isRightKanExtension X n
+  apply Functor.isUniversalOfIsRightKanExtension
+
+variable [∀ (F : (SimplexCategory.Truncated n)ᵒᵖ ⥤ C),
+    (SimplexCategory.Truncated.inclusion n).op.HasRightKanExtension F]
+
+/-- There is a map of costructured arrows from `rightExtensionInclusion X n` to the right extension
+of the `n`-truncation of `X` defined by the counit of `coskAdj n`. -/
+noncomputable def Truncated.rightExtensionCosk.hom : Truncated.rightExtensionInclusion X n ⟶
+    RightExtension.mk _
+      ((coskAdj n).counit.app ((Truncated.inclusion n).op ⋙ X)) :=
+  CostructuredArrow.homMk ((coskAdj n).unit.app X) ((coskAdj n).left_triangle_components X)
+
+instance Truncated.isRightKanExtensionCosk : IsRightKanExtension
+    ((Truncated.cosk n).obj ((inclusion n).op ⋙ X))
+    ((coskAdj n).counit.app ((inclusion n).op ⋙ X)) := by
+  unfold Truncated.cosk coskAdj
+  rw [ranAdjunction_counit]
+  infer_instance
+
+/-- The map `coskRightExtension.hom X` is a natural transformation between two right Kan extensions
+of the diagram `Truncated.inclusion.op ⋙ X` and thus is an isomorphism. -/
+instance IsCoskeletal.coskRightExtension.isIso_hom : IsIso (rightExtensionCosk.hom X n) :=
+  isIso_of_isTerminal (IsCoskeletal.isUniversalOfIsRightKanExtension X n)
+    (((Truncated.cosk n).obj
+      ((Truncated.inclusion n).op ⋙ X)).isUniversalOfIsRightKanExtension
+        ((coskAdj n).counit.app ((Truncated.inclusion n).op ⋙ X)))
+      (rightExtensionCosk.hom X n)
+
+/-- The map `coskRightExtension.hom X` is a natural transformation between two right Kan extensions
+of the diagram `Truncated.inclusion.op ⋙ X` and thus is an isomorphism. -/
+noncomputable def IsCoskeletal.coskRightExtension.homIso :
+    Truncated.rightExtensionInclusion X n ≅ RightExtension.mk _
+      ((coskAdj n).counit.app ((Truncated.inclusion n).op ⋙ X)) :=
+  asIso (rightExtensionCosk.hom X n)
+
+/-- The isomorphism `X ≅ (cosk n).obj X` that exists when `X` is coskeletal and the
+`n`-coskeleton functor exists.-/
+noncomputable def isoCoskOfIsCoskeletal : X ≅ (cosk n).obj X :=
+  (CostructuredArrow.proj ((whiskeringLeft _ _ _).obj (Truncated.inclusion n).op)
+    ((Truncated.inclusion n).op ⋙ X)).mapIso (IsCoskeletal.coskRightExtension.homIso X n)
```

Comment:
A few of the definitions before this one are unnecessary, a better approach would be:
```lean
-- this instance should be in `SimplicialObject.Basic`
instance : ((cosk n).obj X).IsRightKanExtension ((coskAdj n).counit.app _) := by
  dsimp [cosk, coskAdj]
  rw [ranAdjunction_counit]
  infer_instance

/-- The isomorphism `X ≅ (cosk n).obj X` that exists when `X` is `n`-coskeletal and the
`n`-coskeleton functor exists.-/
noncomputable def isoCoskOfIsCoskeletal : X ≅ (cosk n).obj X :=
  X.rightKanExtensionUnique (L := (Truncated.inclusion n).op) (α := 𝟙 _)
    ((cosk n).obj X) ((coskAdj n).counit.app _)

@[simp]
lemma isoCoskOfIsCoskeletal_hom :
    (X.isoCoskOfIsCoskeletal n).hom = (coskAdj n).unit.app X := by
  sorry -- use `hom_ext_of_isRightKanExtension` ?
```

Reply:
Then, using `isRightKanExtension_iff_isIso`, I seems to me it should be easy to show the iff lemma that `X.IsCoskeletal n` iff `IsIso ((coskAdj n).unit.app X)`. If would be nice to add this to this PR.

Reply:
@joelriou I completed the defs you suggested but had a lot of difficulty, see:

* the four instances `Truncated.isRightKanExtensionCosk` with zero to three primes, all of which Lean couldn't infer at some point 
* `isRightKanExtensionCosk_iff_isIso` vs `isCoskeletal_iff_isIso` (here I'm clearly missing something basic).

I get that the terms I built along the way to the original explicit isomorphism in `isoCoskOfIsCoskeletal` were not of general interest to anyone, but I found that approach to be more straightforward. (I get, though, that this is because I don't know my way around the more general theorems in Mathlib.)

Reply:
The fact that you had to construct so many variants of the same instance could be a sign that `cosk` should not be defined as `abbrev`, but as a `def`. Let us consider this out of the scope of this PR.

Reply:
I take this to mean that this is a chore to be explored later in a separate PR. I'll make a note of this.